src/Core/Error.js

```diff
@@ -0,0 +1,78 @@
+/**
+ * @param {Element} element
+ * @param {*} err
+ * @param {string} from
+ * @param {number} signature
+ * @param {Element?}
+ */
+function invokeErrorBoundary (element, err, from, signature) {
+	var error = getErrorException(element, err, from)
+	var snapshot = getErrorElement(element, error, from, signature)
+
+	if (error.report)
```

Comment:
Alternatively this could look like `err.preventDefault()` to silence the default behaviour of logging the error.

```js
class {
	componentDidCatch({stack, message, ...err}) {
		err.preventDefault()
	}
	render() {...}
}
```

vs how you might do this now.

```js
class {
	componentDidCatch({stack, message, ...err}) {
		err.report = false
	}
	render() {...}
}
```

Or it could not log at all in a server environment?

Reply:
I think `preventDefault()` is better. (Though I still think acting like a promise is best, where `.catch()` silences the error unless re-thrown)

I wouldn't change the behavior between server/client, but I do wonder if uncaught should actually `throw` globally... Would basically force a top-level `didCatch()` with logging to prevent server crashes. I like this because the developer **must** consider failure explicitly.

One thing I noticed was that if I accept an error, and return nothing, the element becomes empty (I think). Whereas if uncaught, the element stays valid. What do you think about `undefined/null` returns not causing the component to unmount?

My current use case (was) to have a single top-level `didCatch()`, that simply logs to the console and (if client-side) sends the log to the server. But by catching it breaks the element.

Reply:
>Whereas if uncaught, the element stays valid. What do you think about undefined/null returns not causing the component to unmount?

Do you have a sample reproduction of this, the expected behavior is that it should unmount similar to how React handles it – [error-handling-in-react-16.html](https://facebook.github.io/react/blog/2017/07/26/error-handling-in-react-16.html)

>My current use case (was) to have a single top-level didCatch(), that simply logs to the console and (if client-side) sends the log to the server. But by catching it breaks the element.

Isn't the element already broken at this state? I think the reasons the react team highlight about why they went this route makes sense.

Reply:
Example of uncaught error

```coffee
Root = -> h 'div', 'xxx'
Root.getInitialState = -> throw new Error 'x'
Root.componentWillUnmount = -> console.log 'unmount' # never logged
render Root, $el = document.createElement 'div'
console.log $el
# <div><div>xxx</div></div>
```

Reply:
I understand the desire to not display 'broken' elements.
My use case was for the application to fail as gracefully as possible (with as little error-handling code). That means keeping the DOM intact, so things like links still work, and text is displayed.
I'd rather show users a partially-broken page than a blank screen. But I see arguments both ways

Reply:
Error boundaries can recover, that is return something to display, defining a componentDidCatch stops it's propagation up the tree so you can choose to render an error state for just the components that fail.

For example if you wanted to avoid defining componentDidCatch for all your components you could define a base error class to extend from that handles error state management for components that need it.
. 
```
class BaseErrorComponent {
componentDidCatch() {
return h('h1', 'Error State')
}
}

class A extends BaseErrorComponent {
....
}
```

Since events don't trigger error boundary this shouldn't happen unless something bad went wrong. Do you have examples components of where your case point would apply?

Reply:
my `state` is a stream object that can propagate errors
e.g.
```coffee
Root.componentDidMount = ->
  this.state.stateStream.subscribe (update, err) =>
    if (err)
      return this.state.set err
```